lib/Analysis/ControlDependenceAnalysis.cpp

```diff
@@ -0,0 +1,333 @@
+          traversePostDomTree(&block, block_succ, &funcReg, &postDomTree,
+                              &traversed_nodes);
+
+          // prints the above traversal only in debug mode
+          printPostDomTreeTraversal(traversed_nodes);
+
+          for (size_t i = 0; i < traversed_nodes.size(); i++) {
+            for (size_t j = 0; j < traversed_nodes[i].size(); j++) {
+              // for every node in every path, add block to its control
+              // dependencies
+              mlir::Block *b = traversed_nodes[i][j]->getBlock();
+
+              if (b == &block || b == block_succ)
+                continue; // skip the start and the end nodes because we have
+                          // taken care of them above
+
+              if (std::find(all_control_deps_map[b].begin(),
+                            all_control_deps_map[b].end(),
+                            &block) == all_control_deps_map[b].end()) {
+                all_control_deps_map[b].push_back(&block);
+              }
+            }
+          }
+
+        } else {
+          // All nodes in the post-dominator tree on the path from the
+          // "least_common_anc" to "block_succ" (including "block_succ") should
+          // be control dependent on "block" easy case of block_succ
+          all_control_deps_map[block_succ].push_back(&block);
+
+          // traverse the tree to get all nodes between "least_common_anc" and
+          // "block_succ"
+          llvm::SmallVector<llvm::SmallVector<mlir::DominanceInfoNode *, 4>, 4>
+              traversed_nodes;
+          traversePostDomTree(least_common_anc, block_succ, &funcReg,
+                              &postDomTree, &traversed_nodes);
+
+          // prints the above traversal only in debug mode
+          printPostDomTreeTraversal(traversed_nodes);
+
+          for (size_t i = 0; i < traversed_nodes.size(); i++) {
+            for (size_t j = 0; j < traversed_nodes[i].size(); j++) {
+              // for every node in every path, add block to its control
+              // dependencies
+              mlir::Block *b = traversed_nodes[i][j]->getBlock();
+              if (b == &block || b == block_succ || b == least_common_anc)
+                continue; // skip the start and the end nodes because we have
+                          // taken care of them above
+              if (std::find(all_control_deps_map[b].begin(),
+                            all_control_deps_map[b].end(),
+                            &block) == all_control_deps_map[b].end()) {
+                all_control_deps_map[b].push_back(&block);
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  // up to this point, we have correct direct dependencies that do not include
+  // nested dependencies (i.e., dependencies of a block's dependencies) call the
+  // following function to include nested dependencies
+  addDepsOfDeps(funcOp);
+}
+
+void ControlDependenceAnalysis::traversePostDomTree(
```

Comment:
We could change the `traversePostDomTree` to `enumeratePathsInPostDomTree` (see comments above).